schneiderWallbox/schneiderwallbox.cpp

```diff
@@ -0,0 +1,245 @@
+    });
+    connect(modbusTcpConnection, &SchneiderWallboxModbusTcpConnection::remoteControllerLifeBitChanged, this, [this](quint16 remoteControllerLifeBit){
+        qCDebug(dcSchneiderElectric()) << "Life bit register changed to" << remoteControllerLifeBit;
+        m_recievedLifeBitRegisterValue = remoteControllerLifeBit;
+    });
+    connect(modbusTcpConnection, &SchneiderWallboxModbusTcpConnection::stationIntensityPhaseXChanged, this, [this](double stationIntensityPhaseX){
+        m_stationIntensityPhaseX = stationIntensityPhaseX;
+    });
+    connect(modbusTcpConnection, &SchneiderWallboxModbusTcpConnection::stationIntensityPhase2Changed, this, [this](double stationIntensityPhase2){
+        m_stationIntensityPhase2 = stationIntensityPhase2;
+    });
+    connect(modbusTcpConnection, &SchneiderWallboxModbusTcpConnection::stationIntensityPhase3Changed, this, [this](double stationIntensityPhase3){
+        m_stationIntensityPhase3 = stationIntensityPhase3;
+    });
+    connect(modbusTcpConnection, &SchneiderWallboxModbusTcpConnection::degradedModeChanged, this, [this](quint16 degradedMode){
+        m_degradedMode = degradedMode;
+    });
+}
+
+// When deleting the object, make sure to end remote mode.
+SchneiderWallbox::~SchneiderWallbox()
+{
+    if (m_recievedLifeBitRegisterValue != 2) {
+        qCDebug(dcSchneiderElectric()) << "Terminating remote mode, sending life bit 2.";
+        QModbusReply *reply = m_modbusTcpConnection->setRemoteControllerLifeBit(2);
+        if (!reply) {
+            qCWarning(dcSchneiderElectric()) << "Sending life bit failed because the reply could not be created.";
+            m_errorOccured = true;
+            return;
+        }
+        connect(reply, &QModbusReply::finished, reply, &QModbusReply::deleteLater);
+        connect(reply, &QModbusReply::errorOccurred, this, [reply] (QModbusDevice::Error error){
+            qCWarning(dcSchneiderElectric()) << "Modbus reply error occurred while sending life bit" << error << reply->errorString();
+            emit reply->finished(); // To make sure it will be deleted
+        });
+    }
+
+    qCDebug(dcSchneiderElectric()) << "Deleting SchneiderWallbox object for device with address" << m_modbusTcpConnection->hostAddress();
+    delete m_modbusTcpConnection;
+}
+
+
+SchneiderWallboxModbusTcpConnection *SchneiderWallbox::modbusTcpConnection()
+{
+    return m_modbusTcpConnection;
+}
+
+void SchneiderWallbox::update()
+{
+    if (!m_modbusTcpConnection->connected()) {
+        return;
+    }
+
+    int phaseCount{1};
+    double currentPower = m_stationIntensityPhaseX * 230;
+    if (m_stationIntensityPhase2 > 0.1) {
+        phaseCount++;
+        currentPower += m_stationIntensityPhase2 * 230;
+    }
+    if (m_stationIntensityPhase3 > 0.1) {
+        phaseCount++;
+        currentPower += m_stationIntensityPhase3 * 230;
+    }
+    if (phaseCount != m_phaseCount) {
+        m_phaseCount = phaseCount;
+        emit phaseCountChanged(phaseCount);
+    }
```

Comment:
without having tested it, as I read it, I believe this is slightly problematic:

considering a wallbox and car which both support 3-phase charging:

when it's *not* charging and every phase current is 0, it would result in a phaseCount of 1 and set the state to 1. As soon as it starts charging it would then set it to 3 and set it back to 1 when it's done.

This would cause an optimizer to believe it can only charge on 1 phase while in fact it does 3 phases and make it work with wrong assumptions, forcing it to re-evaluate all of it's output as soon as charging starts.

The way I would write this is:

```
int phaseCount = 0;
if (m_stationIntensityPhase1 > 0.1) phaseCount++;
if (m_stationIntensityPhase2 > 0.1) phaseCount++;
if (m_stationIntensityPhase3 > 0.1) phaseCount++;

// Only update if we actually know the amount of phases
if (phaseCount > 0 && phaseCount != m_phaseCount) {
  m_phaseCount = phaseCount;
  emit ...
}
```

This would cause way less updates to the phasecount state and once it synced up on the wallbox/car combination it would always give the same result until the user really plugs in a different car or uses another cable, which should virtually never happen.